From a46da25b97c4bff044073bf6081d4577ca239d22 Mon Sep 17 00:00:00 2001
From: Aaron Echols <atechols@gmail.com>
Date: Fri, 28 Nov 2014 18:39:26 -0500
Subject: [PATCH 08/17] framework: port IME selector notification toggle (2/2)

luca020400: Cleanup to take the default value in account
            and move to LineageSettings

Change-Id: Ib837fd50cd56e16316c4fe79817987395b38c29f
---
 .../android/server/InputMethodManagerService.java  | 27 ++++++++++++++++++++--
 1 file changed, 25 insertions(+), 2 deletions(-)

diff --git a/services/core/java/com/android/server/InputMethodManagerService.java b/services/core/java/com/android/server/InputMethodManagerService.java
index c23757f..6224824 100644
--- a/services/core/java/com/android/server/InputMethodManagerService.java
+++ b/services/core/java/com/android/server/InputMethodManagerService.java
@@ -165,6 +165,9 @@ import java.util.Locale;
 import java.util.WeakHashMap;
 import java.util.concurrent.atomic.AtomicInteger;
 
+import lineageos.providers.LineageSettings;
+import lineageos.providers.LineageSettings.LineageSettingNotFoundException;
+
 /**
  * This class provides a system service that manages input methods.
  */
@@ -286,6 +289,7 @@ public class InputMethodManagerService extends IInputMethodManager.Stub
     private Notification.Builder mImeSwitcherNotification;
     private PendingIntent mImeSwitchPendingIntent;
     private boolean mShowOngoingImeSwitcherForPhones;
+    private boolean mShowOngoingImeSwitcherForPhonesDefault;
     private boolean mNotificationShown;
 
     static class SessionState {
@@ -779,6 +783,13 @@ public class InputMethodManagerService extends IInputMethodManager.Stub
                     Settings.Secure.SHOW_IME_WITH_HARD_KEYBOARD), false, this, userId);
             resolver.registerContentObserver(Settings.Secure.getUriFor(
                     Settings.Secure.ACCESSIBILITY_SOFT_KEYBOARD_MODE), false, this, userId);
+            resolver.registerContentObserver(LineageSettings.System.getUriFor(
+                    LineageSettings.System.STATUS_BAR_IME_SWITCHER), false,
+                    new ContentObserver(mHandler) {
+                        public void onChange(boolean selfChange) {
+                            updateFromSettingsLocked(true);
+                        }
+                    }, userId);
             mRegistered = true;
         }
 
@@ -1319,6 +1330,7 @@ public class InputMethodManagerService extends IInputMethodManager.Stub
         mImeSwitchPendingIntent = PendingIntent.getBroadcast(mContext, 0, intent, 0);
 
         mShowOngoingImeSwitcherForPhones = false;
+        mShowOngoingImeSwitcherForPhonesDefault = false;
 
         mNotificationShown = false;
         int userId = 0;
@@ -1450,9 +1462,9 @@ public class InputMethodManagerService extends IInputMethodManager.Stub
                     mStatusBar.setIconVisibility(mSlotIme, false);
                 }
                 updateSystemUiLocked(mCurToken, mImeWindowVis, mBackDisposition);
-                mShowOngoingImeSwitcherForPhones = mRes.getBoolean(
+                mShowOngoingImeSwitcherForPhonesDefault = mRes.getBoolean(
                         com.android.internal.R.bool.show_ongoing_ime_switcher);
-                if (mShowOngoingImeSwitcherForPhones) {
+                if (mShowOngoingImeSwitcherForPhonesDefault) {
                     mWindowManagerInternal.setOnHardKeyboardStatusChangeListener(
                             mHardKeyboardListener);
                 }
@@ -2371,6 +2383,17 @@ public class InputMethodManagerService extends IInputMethodManager.Stub
             // There is no longer an input method set, so stop any current one.
             resetCurrentMethodAndClient(InputMethodClient.UNBIND_REASON_NO_IME);
         }
+
+        // Check if phone IME switcher is disabled
+        try {
+            mShowOngoingImeSwitcherForPhones = LineageSettings.System.getInt(
+                    mContext.getContentResolver(),
+                    LineageSettings.System.STATUS_BAR_IME_SWITCHER) == 1 &&
+                    mShowOngoingImeSwitcherForPhonesDefault;
+        } catch (LineageSettingNotFoundException e) {
+            mShowOngoingImeSwitcherForPhones = mShowOngoingImeSwitcherForPhonesDefault;
+        }
+
         // Here is not the perfect place to reset the switching controller. Ideally
         // mSwitchingController and mSettings should be able to share the same state.
         // TODO: Make sure that mSwitchingController and mSettings are sharing the
-- 
2.7.4

